### PR TITLE
Fix incremental build

### LINKED ISSUE #2701: Coreclr Windows Incremental build is broken
Unfortunately the incremental build issue introduced in #2464 was not  resolved by #2494 for windows

### PR DESCRIPTION
**Pull request by ramarag**
This fixes #2701 

@janvorli  @gkhanna79  @jkotas  @schellap  PTAL

### CODE REVIEW

**Comment by ramarag** [comment]
@dotnet-bot test 
FreeBSD x64 Debug Build
FreeBSD x64 Release Build 
OSX x64 Release Build and Test

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,8 @@ project(CoreCLR)
 # Enable @rpath support for shared libraries.
 set(MACOSX_RPATH ON)
 
+#SET_PROPERTY(GLOBAL PROPERTY GLOBAL_DEPENDS_DEBUG_MODE 1)
```
Comment (janvorli) [inline CMakeLists.txt]: Could you please remove this commented out line? I have forgotten to do so, it was just for debugging purposes.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -9,6 +8,42 @@ include_directories("md/inc")
...
+      COMMENT "Generating Eventing Files"
+      OUTPUT ${GENERATED_INCLUDE_DIR}/ClrEtwAll.h
+      OUTPUT ${GENERATED_INCLUDE_DIR}/clretwallmain.h
+      OUTPUT ${GENERATED_INCLUDE_DIR}/etmdummy.h
+      COMMAND ${MC} -h ${GENERATED_INCLUDE_DIR} -r ${GENERATED_INCLUDE_DIR} -b -co -um -p FireEtw "${VM_DIR}/ClrEtwAll.man"
+      COMMAND ${PYTHON} -B -Wall -Werror "${CLR_DIR}/src/scripts/genWinEtw.py" --man "${VM_DIR}/ClrEtwAll.man" --exc "${VM_DIR}/ClrEtwAllMeta.lst" --eventheader "${GENERATED_INCLUDE_DIR}/ClrEtwAll.h" --macroheader "${GENERATED_INCLUDE_DIR}/clretwallmain.h" --dummy "${GENERATED_INCLUDE_DIR}/etmdummy.h"
+      DEPENDS "${CLR_DIR}/src/scripts/genWinEtw.py" "${VM_DIR}/ClrEtwAll.man" "${VM_DIR}/ClrEtwAllMeta.lst"
+    )
+else()
+    add_custom_command(
```
Comment (janvorli) [inline src/CMakeLists.txt]: Please unify these for Unix / win32. They differ only in the python script, so you can just create a CMake variable containing the script path in the if / else and use it in the add_custom_command that you move out.

File: src/dlls/clretwrc/CMakeLists.txt [hunk 3]
```diff
@@ -12,8 +12,7 @@ endif(WIN32)
 
 add_definitions(-DFX_VER_INTERNALNAME_STR=clretwrc.dll)
 
-add_library(clretwrc SHARED
-  clretwrc.rc
+add_library_clr(clretwrc SHARED
```
Comment (janvorli) [inline src/dlls/clretwrc/CMakeLists.txt]: This is missing the clretwrc.rc file, that's why the build in the lab is failing.

File: build.sh [hunk 4]
```diff
@@ -431,9 +446,8 @@ __BinDir="$__RootBinDir/Product/$__BuildOS.$__BuildArch.$__BuildType"
 __PackagesBinDir="$__BinDir/.nuget"
 __ToolsDir="$__RootBinDir/tools"
 __TestWorkingDir="$__RootBinDir/tests/$__BuildOS.$__BuildArch.$__BuildType"
-__IntermediatesDir="$__RootBinDir/obj/$__BuildOS.$__BuildArch.$__BuildType"
+export __IntermediatesDir="$__RootBinDir/obj/$__BuildOS.$__BuildArch.$__BuildType"
```
Comment (janvorli) [inline build.sh]: Why does it need to be exported?

File: build.sh [hunk 5]
```diff
@@ -83,18 +97,19 @@ build_coreclr()
...
         exit
     fi
 
+    rm -rf "$__GeneratedIntermediateEventProvider"
+
     # Do not do any more processing from now on:
     # All set to commence the build
 
-    echo "Commencing build of native components for $__BuildOS.$__BuildArch.$__BuildType in $__IntermediatesDir"
+    echo "Commencing build of native components for $__BuildOS.$__BuildArch.$__BuildType in $__GeneratedIntermediate"
```
Comment (janvorli) [inline build.sh]: It seems this should have stayed unchanged - the components are still built in the __IntermediatesDir, right? Unless I am mistaken, only the python generated stuff is in the __GeneratedIntermediate.

Comment (ramarag) [inline build.sh]: You are correct

File: build.sh (hunk 4 above)
Comment (ramarag) [inline build.sh]: I am using this within the CMake to get to the python generated CMake files

Comment (janvorli) [inline build.sh]: Ah, I see, it is used in the `src/pal/src/CMakeLists.txt`

File: src/CMakeLists.txt (hunk 2 above)
Comment (ramarag) [inline src/CMakeLists.txt]: done

File: src/dlls/clretwrc/CMakeLists.txt (hunk 3 above)
Comment (ramarag) [inline src/dlls/clretwrc/CMakeLists.txt]: done

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -9,6 +8,42 @@ include_directories("md/inc")
 include_directories("classlibnative/bcltype")
 include_directories("classlibnative/cryptography")
 include_directories("classlibnative/inc")
+include_directories("${GENERATED_INCLUDE_DIR}")
+
+set (GeneratingEventFilesCommand)
+if(WIN32)
+    list (APPEND GeneratingEventFilesCommand
+          COMMAND ${MC} -h ${GENERATED_INCLUDE_DIR} -r ${GENERATED_INCLUDE_DIR} -b -co -um -p FireEtw "${VM_DIR}/ClrEtwAll.man"
```
Comment (janvorli) [inline src/CMakeLists.txt]: Ah, I've missed the fact that the MC invocation is another difference in the command. 
So if I may suggest to restructure it a bit differently, I'd prefer it to be this way:
1) Move the MC invocation to a separate command that you would add only for WIN32, let its output be the ClrEtwAll.h (and ClrEtwAll.cpp - not necessary, but for completeness).
2) Put the `COMMAND ${PYTHON} -B -Wall ...` back to the common add_custom_command, parametrizing only the python script
3) Set a variable with the python script path in the if / else
4) Remove the ClrEtwAll.h output from the common custom command

**Comment by janvorli** [comment]
LGTM otherwise

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +8,66 @@ include_directories("md/inc")
...
 include_directories("classlibnative/inc")
+include_directories("${GENERATED_INCLUDE_DIR}")
+
+set (ScriptGeneratedEventFiles
+      ${GENERATED_INCLUDE_DIR}/clretwallmain.h
+      ${GENERATED_INCLUDE_DIR}/etmdummy.h
+    )
+set (GeneratedEventFiles)
+
+if(WIN32)
```
Comment (janvorli) [inline src/CMakeLists.txt]: I am sorry, I guess I must have confused you regarding what I was intending to achieve. What I was after was to make it shorter. The custom target is for handling the header file dependencies, there is no need to use it for the .rc and for the ClrEtwAll.h which is not used anywhere in the codebase as a dependency. So e.g. this:
```cmake
if(WIN32)
  set (GenEventFilesScript "${CLR_DIR}/src/scripts/genWinEtw.py")
    add_custom_command(
      COMMENT "Generating ETW resource Files"
      COMMAND ${MC} -h ${GENERATED_INCLUDE_DIR} -r ${GENERATED_INCLUDE_DIR} -b -co -um -p FireEtw "${VM_DIR}/ClrEtwAll.man"
      OUTPUT ${GENERATED_INCLUDE_DIR}/ClrEtwAll.h
      OUTPUT ${GENERATED_INCLUDE_DIR}/ClrEtwAll.rc
      DEPENDS "${VM_DIR}/ClrEtwAll.man"
       )
else()
  set (GenEventFilesScript "${CLR_DIR}/src/scripts/genXplatEventing.py")
endif()

set (ScriptGeneratedEventFiles
      ${GENERATED_INCLUDE_DIR}/clretwallmain.h
      ${GENERATED_INCLUDE_DIR}/etmdummy.h
    )

add_custom_command(
  COMMENT "Generating Eventing Files"
  COMMAND ${PYTHON} -B -Wall -Werror ${GenEventFilesScript} ${GenEventArgs} --man "${VM_DIR}/ClrEtwAll.man" --exc "${VM_DIR}/ClrEtwAllMeta.lst" --dummy "${GENERATED_INCLUDE_DIR}/etmdummy.h"
  OUTPUT ${ScriptGeneratedEventFiles}
  DEPENDS ${GenEventFilesScript} "${VM_DIR}/ClrEtwAll.man" "${VM_DIR}/ClrEtwAllMeta.lst" "${CLR_DIR}/src/scripts/genXplatEventing.py"
    )

add_custom_target(
  GeneratedEventingFiles
  DEPENDS ${ScriptGeneratedEventFiles}
)

```

Comment (janvorli) [inline src/CMakeLists.txt]: I am sorry, please disregard this comment. I have missed the fact that the generated clretwallmain.h includes the generated ClrEtwAll.h.

Comment (ramarag) [inline src/CMakeLists.txt]: I have added comments in the CMake file to callout the dependecies

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -9,6 +8,74 @@ include_directories("md/inc")
...
 include_directories("classlibnative/inc")
+include_directories("${GENERATED_INCLUDE_DIR}")
+
+# The Following Logic is used to wire up Build dependencies for Generated files in Event Logging 
+# genXplatEventing.py            - has the core logic for parsing Event Schema
+# genWinEtw.py                   - Uses genXplatEventing to generate Windows Specific ETW Files
+# clretwallmain.h and etmdummy.h - Provides the Event Logging Functionality to the VM
+# clrxplatevents.h               - Used by clretwallmain.h on Non Windows platform
+# ClrEtwAll.h                    - Used by clretwallmain.h on  Windows 
+# ClrEtwAll.rc                   - Used by src/dlls/clretwrc/clretrw.rc on Windows
```
Comment (janvorli) [inline src/CMakeLists.txt]: Great, very nice!

**Comment by janvorli** [comment]
LGTM. Btw, it seems you'll need to resolve some conflicts

**Comment by ramarag** [comment]
@dotnet-bot test 
OSX x64 Release Build and Test
 FreeBSD x64 Release Build

**Pull request merged**